commit 8d9370864ffaf75de3e39e3fc013b8fa955d944c (refs/changes/54/854/8)
Author: Richard Bradley <richard.bradley@joyent.com>
Date:   2016-11-16T09:49:23+00:00 (2 years, 11 months ago)
    
    MANTA-2587 manta-net.sh Needs to Support Adding tags to Aggregations
    MANTA-2150 manta net config and validate allows for mac addresses to have single-digits between ::

diff --git a/Makefile b/Makefile
index 37369b3..74b9d2f 100644
--- a/Makefile
+++ b/Makefile
@@ -43,7 +43,7 @@ BASH_FILES	 = scripts/user-script.sh  \
 		   bin/manta-deploy-lab \
 		   networking/manta-net.sh
 DOC_FILES	 = index.md
-JS_FILES	:= $(shell find cmd lib test -name '*.js')
+JS_FILES	:= $(shell find cmd lib test networking -name '*.js')
 JSL_FILES_NODE	 = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
 JSON_FILES	 = package.json \
diff --git a/networking/manta-net.sh b/networking/manta-net.sh
index 708c774..35aa049 100755
--- a/networking/manta-net.sh
+++ b/networking/manta-net.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2016, Joyent, Inc.
 #
 
 #
@@ -376,7 +376,7 @@ function fetch_ufds_ids
 
 function add_tags
 {
-	local stanza nid nodes n map nmap mac ouuid
+	local stanza nid nodes tag n map if_type interface
 	stanza=$1
 	nid=$2
 	[[ -z "$stanza" ]] && fatal "missing required stanza for add_tags"
@@ -388,25 +388,113 @@ function add_tags
 	[[ $? -eq 0 ]] || fatal "failed to get nic tag for stanza $stanza"
 	[[ -z "$tag" ]] && fatal "unexpected empty nic tag"
 	for n in $nodes; do
-		map=$(json mac_mappings.$n.$tag < $mn_config)
-		[[ $? -eq 0 ]] || fatal "failed to get mac mapping via json"
-		[[ -z "$map" ]] && "empty mac mapping"
-		mac=$(echo $map | sed -e 's/^0:/00:/' -e 's/:0:/:00:/g' -e \
-		    's/:0$/:00/g')
-		[[ $? -eq 0 ]] || fatal "failed to translate mac"
-		nmap=$(echo $mac | sed 's/://g')
-		[[ $? -eq 0 ]] || fatal "failed to translate mac"
-		ouuid=$(sdc-napi /nics/$nmap | json -H belongs_to_uuid)
-		[[ $? -eq 0 ]] || fatal "failed to get server uuid for: $map"
-		[[ "$ouuid" == "$n" ]] || fatal "mapping does nic not match " \
-		    "nic owner for $map, expected it to be $n, found $ouuid"
-		sdc-napi /nics/$nmap | json -H nic_tags_provided | json -a | \
-		    grep -q "^$tag$" && continue
-		sdc-server update-nictags -s $n "${tag}_nic=$mac"
-		[[ $? -eq 0 ]] || fatal "failed to add nic tag"
+		map=$(json mac_mappings.$n.$tag \
+		    nic_mappings.$n.$tag < $mn_config)
+		[[ $? -eq 0 ]] || fatal "failed to get nic mapping via json"
+		[[ -z "$map" ]] && fatal "empty $tag interface mapping " \
+		    "for $n"
+
+		# validate.js enforces $map to either be a JSON String or
+		# Object, which represents the old and new style, respectively,
+		# of defining what nics should be tagged.
+		if echo $map | json --validate -q; then
+			if_type=$(echo $map | json -ka)
+			[[ $? -eq 0 ]] || fatal "failed to translate mapping"
+			[[ -z "$if_type" ]] && fatal "unexpected empty " \
+			    "interface type for nic_mappings.$n.$tag"
+			interface=$(echo $map | json $if_type)
+			[[ $? -eq 0 ]] || fatal "failed to translate mapping"
+			[[ -z "$interface" ]] && fatal "unexpected empty " \
+			    "interface value for nic_mappings.$n.$tag"
+		else
+			if_type="mac"
+			interface=$map
+		fi
+
+		if [[ "$if_type" == "mac" ]]; then
+			add_tag_to_mac $interface $n $tag
+		elif [[ "$if_type" == "aggr" ]]; then
+			add_tag_to_aggr $interface $n $tag
+		fi
 	done
 }
 
+function add_tag_to_mac
+{
+	local interface node tag mac nmap nres ouuid
+	interface=$1
+	node=$2
+	tag=$3
+	[[ -z "$interface" ]] && fatal "missing required interface " \
+	    "for add_tag_to_mac"
+	[[ -z "$node" ]] && fatal "missing required node for add_tag_to_mac"
+	[[ -z "$tag" ]] && fatal "missing required tag for add_tag_to_mac"
+	mac=$(echo $interface | sed -e 's/\<.\>/0&/g')
+	[[ $? -eq 0 ]] || fatal "failed to translate mac"
+	[[ -z "$mac" ]] && fatal "unexpected empty mac address after " \
+	    "translation from $interface"
+	nmap=$(echo $mac | sed -e 's/://g')
+	[[ $? -eq 0 ]] || fatal "failed to translate mac"
+	[[ -z "$nmap" ]] && fatal "unexpected empty napi identifier after " \
+	    "translation from $mac"
+	nres=$(sdc-napi /nics/$nmap | json -H)
+	[[ $? -eq 0 ]] || fatal "unexpected failure reaching napi"
+	# we don't check for an empty $nres here because napi may respond with
+	# a 404, but also a "not found"-type object. We check if we have a 200
+	# on a nic later by checking for belongs_to_uuid in the response
+	ouuid=$(echo $nres | json belongs_to_uuid)
+	[[ $? -eq 0 ]] || fatal "failed to get server uuid for $interface"
+	# we use the existence of belongs_to_uuid in the napi response to tell
+	# if we've found a nic or not. if not, our message will contain enough
+	# information for an operator to manually confirm
+	[[ -z "$ouuid" ]] && fatal "failed to get nic details" \
+	    "for $interface at /nics/$nmap"
+	[[ "$ouuid" == "$n" ]] || fatal "mapping does not match nic owner" \
+	    "for $interface, expected it to be $n, found $ouuid"
+	echo $nres | json nic_tags_provided | json -a | \
+	    grep -q "^$tag$" && continue
+	sdc-server update-nictags -s $n "${tag}_nic=$mac"
+	[[ $? -eq 0 ]] || fatal "failed to add nic tag"
+}
+
+function add_tag_to_aggr
+{
+	local interface node tag aggr nres ouuid existing_tags update_tags
+	interface=$1
+	node=$2
+	tag=$3
+	[[ -z "$interface" ]] && fatal "missing required interface " \
+	    "for add_tag_to_aggr"
+	[[ -z "$node" ]] && fatal "missing required node for add_tag_to_aggr"
+	[[ -z "$tag" ]] && fatal "missing required tag for add_tag_to_aggr"
+	aggr=$n-$interface
+	nres=$(sdc-napi /aggregations/$aggr | json -H)
+	[[ $? -eq 0 ]] || fatal "unexpected failure reaching napi"
+	ouuid=$(echo $nres | json belongs_to_uuid)
+	[[ $? -eq 0 ]] || fatal "failed to get server uuid for $aggr"
+	[[ -z "$ouuid" ]] && fatal "failed to get aggr details" \
+	    "for $interface at /aggregations/$aggr"
+	[[ "$ouuid" == "$n" ]] || fatal "mapping does not match aggr owner " \
+	    "for $aggr, expected it to be $n, found $ouuid"
+	echo $nres | json nic_tags_provided | json -a | \
+	    grep -q "^$tag$" && continue
+	existing_tags=$(echo $nres | json -a nic_tags_provided)
+	[[ $? -eq 0 ]] || fatal "failed to get existing nic tags for $aggr"
+	update_tags=$(echo $existing_tags $'\n' "[\"$tag\"]" | json -g)
+	[[ $? -eq 0 ]] || fatal "failed to merge nic tags for $aggr"
+	[[ -z "$update_tags" ]] && fatal "unexpected empty updated nic tags " \
+	    "list for $aggr"
+	# the following will update the aggregation, but we pipe the response
+	# directly into `json` to confirm the response contains the requested
+	# nic tag. The final conditional will protect us from non-0 exits from
+	# sdc-napi, as well as any other napi http error, in which case our
+	# `grep` will not match and exit non-0
+	sdc-napi /aggregations/$aggr -X PUT -d "{
+		\"nic_tags_provided\": $update_tags
+	}" | json -H nic_tags_provided | json -a | \
+	    grep -q "^$tag$" || fatal "failed to add nic tag"
+}
+
 function append_routes
 {
 	local routes outfile
diff --git a/networking/validate.js b/networking/validate.js
index 3adfbf9..b749a6a 100644
--- a/networking/validate.js
+++ b/networking/validate.js
@@ -19,9 +19,37 @@ function fatal(message)
 	process.exit(1);
 }
 
+function validateMappings(network, mappings, spec)
+{
+	var i, node, tag, nic;
+	for (i = 0; i < spec[network + '_nodes'].length; i++) {
+		node = spec[network + '_nodes'][i];
+		tag = spec[network]['nic_tag'];
+		if (!(node in mappings))
+			fatal('missing node from mappings: ' + node);
+		if (!(tag in mappings[node]))
+			fatal('missing tag ' + tag + ' in mappings.' + node);
+		nic = mappings[node][tag];
+		if ('mac_mappings' in spec) {
+			if (!(typeof (nic) === 'string'))
+				fatal('mac_mappings.' + node + '.' + tag +
+				    ' must be a string');
+		}
+		if ('nic_mappings' in spec) {
+			if (!(typeof (nic) === 'object'))
+				fatal('nic_mappings.' + node + '.' + tag +
+				    ' must be an object');
+			if (Object.keys(nic).length > 1 ||
+			    !('mac' in nic) && !('aggr' in nic))
+				fatal('nic_mappings.' + node + '.' + tag +
+				    ' must only have property "mac" or "aggr"');
+		}
+	}
+}
+
 function main()
 {
-	var spec, az, i;
+	var spec, i, az, mappings;
 
 	if (process.argv.length !== 3) {
 		console.error('validate.js: <file>');
@@ -29,9 +57,6 @@ function main()
 	}
 	spec = JSON.parse(mod_fs.readFileSync(process.argv[2]));
 
-	if (!('mac_mappings' in spec))
-		fatal('missing mac_mappings');
-
 	if (!('marlin_nodes' in spec))
 		fatal('missing marlin_nodes');
 
@@ -41,17 +66,6 @@ function main()
 	if (spec['marlin_nodes'].length === 0)
 		fatal('no marlin CNs listed');
 
-	for (i = 0; i < spec['marlin_nodes'].length; i++) {
-		if (!(spec['marlin_nodes'][i] in spec['mac_mappings']))
-			fatal('missing node from mac_mappings: ' +
-			    spec['marlin_nodes'][i]);
-		if (!(spec['marlin']['nic_tag'] in
-		    spec['mac_mappings'][spec['marlin_nodes'][i]])) {
-			fatal('missing tag ' + spec['marlin']['nic_tag'] +
-			    ' in mac_mappings.' + spec['marlin_nodes'][i]);
-		}
-	}
-
 	if (!('manta_nodes' in spec))
 		fatal('missing manta_nodes');
 
@@ -61,17 +75,6 @@ function main()
 	if (spec['manta_nodes'].length === 0)
 		fatal('no indexing CNs listed');
 
-	for (i = 0; i < spec['manta_nodes'].length; i++) {
-		if (!(spec['manta_nodes'][i] in spec['mac_mappings']))
-			fatal('missing node from mac_mappings: ' +
-			    spec['manta_nodes'][i]);
-		if (!(spec['manta']['nic_tag'] in
-		    spec['mac_mappings'][spec['manta_nodes'][i]])) {
-			fatal('missing tag ' + spec['manta']['nic_tag'] +
-			    ' in mac_mappings.' + spec['manta_nodes'][i]);
-		}
-	}
-
 	if (!('azs' in spec))
 		fatal('missing availability zone list');
 
@@ -108,6 +111,19 @@ function main()
 	if (!('network' in spec['marlin']))
 		fatal('marlin section missing network name');
 
+	if (!('mac_mappings' in spec) &&
+	    !('nic_mappings' in spec))
+		fatal('missing nic_mappings or mac_mappings');
+
+	if (('mac_mappings' in spec) &&
+	    ('nic_mappings' in spec))
+		fatal('either nic_mappings or mac_mappings');
+
+	mappings = spec['mac_mappings'] || spec['nic_mappings'];
+
+	validateMappings('manta', mappings, spec);
+	validateMappings('marlin', mappings, spec);
+
 	for (i = 0; i < spec['azs'].length; i++) {
 		az = spec['azs'][i];
 		if (!(az in spec['admin']))
